### PR TITLE
Memory reorganizations

### PR DESCRIPTION
**Pull request by mpeychev**

### CODE REVIEW

**Comment by mpeychev** [comment]
Ignore the fact that the branch is called `simd`.

**Comment by ericpts** [comment]
I am sort of hesitant of directly approving, because I do not like the way of indexing using 
`k * nstates + j`. It just completely seems like something the compiler should be doing instead of us, and that perhaps there should be a cleaner way which preserves the nice accessing pattern (i.e. `mat[i][j]`), while maintaining the performance.

**Comment by ericpts** [comment]
I am sort of hesitant of directly approving, because I do not like the way of indexing using 
`k * nstates + j`. It just completely seems like something the compiler should be doing instead of us, and that perhaps there should be a cleaner way which preserves the nice accessing pattern (i.e. `mat[i][j]`), while maintaining the performance.

**Comment by ericpts** [comment]
i.e., something like this maybe https://www.boost.org/doc/libs/1_38_0/libs/numeric/ublas/doc/matrix.htm#matrix

**Comment by ericpts** [comment]
Or I guess the main point of going to direct accesses like `j * nstates + i` is to make loading/storing memory more straight-forward, as opposed to something hidden behind an API.

File: src/baumwelch_mem_reorg.cpp [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    for (int t = time - 2; t >= 0; t--) {
+        for (int i = 0; i < nstates; ++i) {
+            double accum = 0.0;
+            for (int j = 0; j < nstates; ++j) {
+                accum += trans_prob[i * nstates + j] *
+                    emit_prob[emissions[t + 1] * nstates + j] *
+                    beta_hat[(t + 1) * nstates + j];
+            }
+            beta_dot[t * nstates + i] = accum;
+            beta_hat[t * nstates + i] = c[t] * accum;
```
Comment (ericpts) [inline src/baumwelch_mem_reorg.cpp]: There are no asserts here, to check that the values are finite.

File: src/baumwelch_mem_reorg.cpp [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        alpha_hat[i] = c[0] * alpha_dot[i];
+    }
+
+    for (int t = 1; t < time; ++t) {
+        Z = 0.0;
+        for (int i = 0; i < nstates; ++i) {
+            alpha_dot[t * nstates + i] = 0.0;
+            for (int j = 0; j < nstates; ++j) {
+                alpha_dot[t * nstates + i] +=
+                    alpha_hat[(t - 1) * nstates + j] * trans_prob[j * nstates + i] *
```
Comment (ericpts) [inline src/baumwelch_mem_reorg.cpp]: Shouldn't that be `(t - 1) * time`?

**Comment by mpeychev** [comment]
The main reason I did it this way is to avoid working with double pointers (i.e. `double **ma3x`).
Because I don't like pointers on the first place 😅 

I agree it is a bit difficult to read, but I wouldn't hide it behind a deep API neither.

Another thing I would like to point is the fact I am using `posix_memalign`. I think the "standard" C++17 way is using `aligned_alloc` but I couldn't make it compile (didn't try too hard tho)

**Comment by mpeychev** [comment]
And yes, I did it this way in preparation to vectorizing the implementation.

Will be merging, feel free to make any changes. I feel like the memory pattern here should be better than previous implementation and maybe can still benefit from further blocking. Haven't examined the code w.r.t. caches in greater detail.

**Comment by ericpts** [comment]
My stats were done using `valgrind --tool=cachegrind`.

In particular, note that memreorg performs `4,196,945,550` instructions, whereas naivescaled performs `10,315,095,242` instructions.
Since they are supposed to be doing the exact same thing, it seems to me that something must be wrong.

**Comment by mpeychev** [comment]
The ops counts (as defined in our cost measure) are the same. I reorganized the last cycles in the end of `IterOnce` function. There is no longer `if (emissions[t] == k)` or sth like that.

This may make some effect on the number of instructions executed.

I am not familiar with valgrind but I think it is indeed likely to be more cache friendly.

File: src/baumwelch_mem_reorg.cpp (hunk 2 above)
Comment (mpeychev) [reply src/baumwelch_mem_reorg.cpp]: I know it is confusing (sorry), but no. You can suppose `time` is the number of rows (height) and `nstates` is the number of columns (width) of the matrix/2D array.

File: src/baumwelch_mem_reorg.cpp (hunk 1 above)
Comment (mpeychev) [reply src/baumwelch_mem_reorg.cpp]: Added in subsequent commit to `master`.

**Pull request merged**